Fix memory usage of PullParser in D2 builds

mihails-strasuns-sociomantic opened the pull request:
Alternative approach to https://github.com/sociomantic-tsunami/ocean/pull/347

Now `PullParser` operates strictly on slices and does no allocations and
`Document` copies those slices into nodes when necessary.

mihails-strasuns-sociomantic committed:
No alloc test in XML PullParser

mihails-strasuns-sociomantic committed:
Remove unnecessary module DDOC sections

mihails-strasuns-sociomantic committed:
Reindent XML PullParser

gavin-norman-sociomantic commented on src/ocean/text/xml/PullParser.d:
```diff
@@ -1,17 +1,13 @@
-
-        Version: Initial release: February 2008
-
-        Authors: Aaron, Kris
```
I thought we'd deliberately left in the tango authors info?

mihails-strasuns-sociomantic replied:
I don't remember details but I remember that SVN logs has shown that real authors are quite different from whatever is left in DDOC. And git history is still present anyway. @leandro-lucarella-sociomantic do you know how important is this part legal-wise?

mihails-strasuns-sociomantic replied:
(restored just in case)

gavin-norman-sociomantic commented on src/ocean/text/xml/PullParser.d:
```diff
@@ -680,10 +670,10 @@ class PullParser(Ch = char)
-    package size_t  len;
-    package Ch[]    text;
-    package Ch*     point;
+    package Ch*      end;
```
Should go in the previous commit.

gavin-norman-sociomantic reviewed (commented):
I'm not really familiar with the XML parser, so won't comment on the details unless there is a need for an extra opinion, in addition to Ben / Don.

don-clugston-sociomantic commented:
I think the approach you've taken here is basically right. I don't think it ever makes sense to have a parser that gives mutable slices. The changes to `Document` do mean that it allocates a lot more memory, and does a lot more copying -- every tag in the document will be copied -- but since it allocates one node for every XML element anyway I don't think it's really a problem.
I've never used `Document` but the `PullParser` changes work for me. I will close my PRs on this topic.

I just have one comment, and it's optional.

Since you are cleaning up the formatting of the code, I think you should also delete the untested code in the undocumented `version` statements. There's this horrendous one in Document.d
```d
version(d)
{
                freelists = 0;          // needed to align the codegen!
}
```
I don't know if this is a typo of `version(discrete)` or if they really had `version=d;` (!)

ben-palmer-sociomantic reviewed (approved):
I agree with Dons comments. I also tested with my code and found no problems.

mihails-strasuns-sociomantic committed:
Fix memory usage of PullParser in D2 builds

Now `PullParser` operates strictly on slices and does no allocations and
`Document` copies those slices into nodes when necessary.

mihails-strasuns-sociomantic committed:
Remove dead code in XML Document

mihails-strasuns-sociomantic commented:
Updated. I must say, amount of `version` code that is likely never used in practice worries me a bit.

don-clugston-sociomantic reviewed (approved):
LGTM

nemanja-boric-sociomantic reviewed (approved):
LGTM
